### DISCUSSION TITLE
[RFC] Internal design documentation about xmake

### DISCUSSION THREAD

**Discussion by inclyc** [xmake-io/xmake#2907] [Ideas]
### Is your feature request related to a problem? Please describe.

At present, xmake does not have an internal design document and ideas to facilitate the energetic developers to contribute code. Maybe we need something like this.

目前 xmake 还没有一个内部的设计文档和思路方便有精力的开发者贡献代码。或许我们需要一个这样的东西

### Describe the solution you'd like

Open a new GitHub repository or create a new directory within this repo for storing documents, and write the function and implementation principle of each component. Contains design decisions, design overview, etc.

新开一个仓库或就在本仓库内部建立存放文档的目录，写入每个部件的功能和实现的原理。包含设计决策、设计概述等。

### Describe alternatives you've considered

None

### Additional context

Now most famous open source projects have internal design documents, which is very important for new community contributors. Some developers may have time to maintain `xmake` (for example some subsystems) but have no time to get familiar with the `xmake` codebase, and providing an internal reference document can be a good solution to this problem.

现在绝大多数著名的开源项目都有内部的设计文档，这对于新社区贡献者而言是十分重要的，可能有人有精力维护`xmake`但没有足够的时间去熟悉`xmake`源代码，而提供一个内部参考文档就能很好的解决这个问题。

CC @waruqi

**Comment by inclyc** [comment]
FWIW in LLVM Project we have [CFE Internals Manual](https://clang.llvm.org/docs/InternalsManual.html)

**Comment by waruqi** [comment]
有一篇老的，https://tboox.org/cn/2017/09/28/xmake-sourcecode-arch/

这种太费精力，暂时没时间搞。
